### PR TITLE
job runners: document the handler interface

### PR DESCRIPTION
**Pull request by oliver-sanders**
> **Sibling:** https://github.com/cylc/cylc-doc/pull/594

The documentation for the job runner handler interface is poor but I found a docstring in cylc-flow with valuable information which wasn't being auto-documented.

* Move existing documentation into a class for easier auto-documentation.
* Pull in a little content from cylc-doc and attempt to pad things out a little.
* Pull in lists of fields for `job_conf` and `submit_opts` obtained from the codebase.

I'm not taking the time for a full documentation re-write right now (no time) so this is mostly just what was already there reformatted and auto documented. The important thing is to expose this documentation as there are people trying to write handlers ATM.

@hjoliver if you get a chance to skim this that would be great.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at https://github.com/cylc/cylc-doc/pull/594
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
@wxtim can you check that the two branches build and that it vaguely makes sense.

File: cylc/flow/job_runner_handlers/documentation.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    List,
+    Tuple,
+)
+
+
+class ExampleHandler():
+    """Documentation for writing job runner handlers.
+
+    Cylc can submit jobs to a number of different job runners
+    (aka batch systems) e.g. Spice and PBS. For a list of built-in integrations
```
Comment (hjoliver) [inline cylc/flow/job_runner_handlers/documentation.py:37]:
```suggestion
    (aka batch systems) e.g. Slurm and PBS. For a list of built-in integrations
```
("Spice"?)

**Review by hjoliver — COMMENTED** [review]
(Looks good from a quick read-through)

File: cylc/flow/job_runner_handlers/documentation.py (hunk 1 above)
Comment (oliver-sanders) [reply cylc/flow/job_runner_handlers/documentation.py:37]: Ah jeez, good spot.

(Spice is the name of an internal system which uses Slurm)

File: cylc/flow/job_runner_handlers/documentation.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+       * ``job_runner_cmd_tmpl``
+
+
+    .. rubric:: An Example
+    
+    The following ``qsub.py`` module overrides the built-in *pbs*
+    job runner handler to change the directive prefix from ``#PBS`` to
+    ``#QSUB``:
+    
+    .. TODO - double check that this still works, it's been a while
```
Comment (wxtim) [inline cylc/flow/job_runner_handlers/documentation.py:142]: TODO

File: cylc/flow/job_runner_handlers/documentation.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        raise NotImplementedError()
+
+    def submit(self, job_file_path: str, submit_opts: dict) -> Tuple[int, str, str]:
+        """Submit a job.
+
+        Submit a job and return an instance of the Popen object for the
+        submission. This method is useful if the job submission requires logic
+        beyond just running a system or shell command.
+
+        See also :py:attr:`ExampleHandler.SUBMIT_CMD`.
```
Comment (wxtim) [inline cylc/flow/job_runner_handlers/documentation.py:386]: This isn't present in ExampleHandler so this reference doesn't build.

**Review by wxtim — COMMENTED** [review]
I'm getting a failure to build with 

``` 
/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/job_runner_handlers/documentation.py:docstring of cylc.flow.job_runner_handlers.documentation.ExampleHandler.REC_ID_FROM_SUBMIT_ERR:1: WARNING: py:class reference target not found: re.Pattern
/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/job_runner_handlers/documentation.py:docstring of cylc.flow.job_runner_handlers.documentation.ExampleHandler.REC_ID_FROM_SUBMIT_OUT:1: WARNING: py:class reference target not found: re.Pattern
/net/home/h02/tpilling/metomi/cylc-flow/cylc/flow/job_runner_handlers/documentation.py:docstring of cylc.flow.job_runner_handlers.documentation.ExampleHandler:3: WARNING: py:attr reference target not found: ExampleHandler.SUBMIT_CMD


```

File: cylc/flow/job_runner_handlers/documentation.py (hunk 2 above)
Comment (oliver-sanders) [reply cylc/flow/job_runner_handlers/documentation.py:142]: Pre-existing TODO.

File: cylc/flow/job_runner_handlers/documentation.py (hunk 3 above)
Comment (oliver-sanders) [reply cylc/flow/job_runner_handlers/documentation.py:386]: I'm guessing this was meant to be `SUBMIT_CMD_TMPL`?

**Comment by oliver-sanders** [comment]
The error on the `re.Pattern` class is because Python has not included this in their docs so it is failing on a broken reference.

Added an ignore in this commit:

https://github.com/cylc/cylc-doc/pull/594/commits/d84e3b7266e3606163e213bb4a6e85151e896bd7

File: cylc/flow/job_runner_handlers/documentation.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+            lines
+
+        """
+        raise NotImplementedError()
+
+    def get_poll_many_cmd(self, job_id_list: List[str]) -> List[str]:
+        """Return a command to poll the specified jobs.
+
+        Args:
+            job_id_list: The listof job ids to poll.
```
Comment (wxtim) [inline cylc/flow/job_runner_handlers/documentation.py:335]:
```suggestion
            job_id_list: The list of job ids to poll.

File: cylc/flow/job_runner_handlers/documentation.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    .. rubric:: Usage
+
+    You can then define a Cylc platform using the handler:
+
+    .. code-block:: cylc
+
+       [platforms]
+           [[my_platform]]
+               job runner = my_handler  # note matches Python module name
+               hosts = localhost
```
Comment (hjoliver) [inline cylc/flow/job_runner_handlers/documentation.py:81]: Maybe indicate that this is a global config fragment.

File: cylc/flow/job_runner_handlers/documentation.py [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+    And configure tasks to submit to it:
+
+    .. code-block:: cylc
+
+       [runtime]
+           [[my_task]]
+               script = echo "Hello World!"
+               platform = my_platform
```
Comment (hjoliver) [inline cylc/flow/job_runner_handlers/documentation.py:90]: And this a workflow config fragment.

File: cylc/flow/job_runner_handlers/documentation.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+    SHOULD_POLL_PROC_GROUP: bool
+    """Poll jobs by PID.
+
+    A boolean to indicate whether it is necessary to poll a job by its PID
+    as well as the job ID.
+
+    """
+
+    REC_ID_FROM_SUBMIT_OUT: re.Pattern
+    """Regular expression to extract job ids from submission stderr.
```
Comment (hjoliver) [inline cylc/flow/job_runner_handlers/documentation.py:246]:
```suggestion
    """Regular expression to extract job IDs from submission stderr.
```

**Review by hjoliver — APPROVED** [review]
Docs build OK for me. Looks good. A few minor suggestions.

File: cylc/flow/job_runner_handlers/documentation.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+       JOB_RUNNER_HANDLER = MyHandler()
+
+    Each job runner handler class should instantiate with no argument.
+
+
+    .. rubric:: Usage
+
+    You can then define a Cylc platform using the handler:
+
+    .. code-block:: cylc
```
Comment (oliver-sanders) [inline cylc/flow/job_runner_handlers/documentation.py:76]:
```suggestion
    .. code-block:: cylc
       :caption: global.cylc
```

File: cylc/flow/job_runner_handlers/documentation.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+
+       [platforms]
+           [[my_platform]]
+               job runner = my_handler  # note matches Python module name
+               hosts = localhost
+
+    And configure tasks to submit to it:
+
+    .. code-block:: cylc
```
Comment (oliver-sanders) [inline cylc/flow/job_runner_handlers/documentation.py:85]:
```suggestion
    .. code-block:: cylc
       :caption: flow.cylc
```

File: cylc/flow/job_runner_handlers/documentation.py [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+
+        Filter the standard output and standard error of the job submission
+        command. This is useful if the job submission command returns
+        information that should just be ignored.
+
+        See also :py:meth:`ExampleHandler.SUBMIT_CMD_TMPL`.
+
+        Args:
+            out: Job submit stdout.
+            err: Job submit stderr.
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/job_runner_handlers/documentation.py:308]: Dunno, this is valid google-style-docstring.

File: cylc/flow/job_runner_handlers/documentation.py (hunk 4 above)
Comment (oliver-sanders) [reply cylc/flow/job_runner_handlers/documentation.py:335]:
```suggestion
            job_id_list: The list of job IDs to poll.
```

**Review by wxtim — APPROVED** [review]

**Pull request merged**